packages/decoder/src/array.ts

```diff
@@ -0,0 +1,52 @@
+import type { ParseOptions } from "@effect/schema/AST"
+import * as PR from "@effect/schema/ParseResult"
+import { Effect, pipe, ReadonlyArray as RA } from "effect"
+
+import { compose } from "@typed/decoder/compose"
+import type { Decoder } from "@typed/decoder/decoder"
+
+export const unknownArray: Decoder<unknown, ReadonlyArray<unknown>> = (i) =>
+  Array.isArray(i) ? PR.success(i) : PR.failure(PR.unexpected(i))
+
+export const fromArray =
+  <I, O>(member: Decoder<I, O>): Decoder<ReadonlyArray<I>, ReadonlyArray<O>> =>
+  (i: ReadonlyArray<I>, options?: ParseOptions) => {
+    return Effect.gen(function*($) {
+      const results = yield* $(
+        Effect.all(
+          pipe(
+            i,
+            RA.map((ix, idx) =>
+              pipe(
+                member(ix, options),
+                Effect.mapError((e: PR.ParseError) => PR.index(idx, e.errors)),
+                Effect.either
+              )
+            )
+          )
+        )
+      )
+      const [failures, successes] = RA.separate(results)
+
+      if (RA.isNonEmptyReadonlyArray(failures)) {
+        return yield* $(Effect.fail(PR.parseError(failures)))
+      }
+
+      return successes
+    })
+  }
+
+export const array = <O>(member: Decoder<unknown, O>): Decoder<unknown, ReadonlyArray<O>> =>
+  pipe(unknownArray, compose(fromArray(member)))
+
+export const fromNonEmptyArray =
+  <I, O>(member: Decoder<I, O>): Decoder<ReadonlyArray<I>, ReadonlyArray<O>> => (i, options) => {
```

Comment:
🔧 Should we update the types for `fromNonEmptyArray` and `nonEmptyArray` to return a `ReadonlyNonEmptyArray<O>`?

Reply:
Yes, the end-developer would like to be able to use `headNonEmpty`, `tailNonEmpty`, etc if it's already been validated by the decoder.

Is there a better way of doing this than using a type assertion?
```
export const fromNonEmptyArray =
  <I, O>(member: Decoder<I, O>): Decoder<ReadonlyArray<I>, NonEmptyReadonlyArray<O>> => (i, options) => {
    if (i.length === 0) {
      return PR.failure(PR.unexpected(i))
    }

    return fromArray(member)(i, options) as PR.ParseResult<NonEmptyReadonlyArray<O>>
  }
```

Reply:
At some level we either need a type-cast, or a type-assertion that makes it "safe" (which is really just a type-cast that's better supported at runtime), so I don't mind the type-cast here

Reply:
Thx, I've added the change.